File: sdks/nodejs/src/request.js
/**
 * Download request with fluent API.
 */

const { spawn } = require('child_process');

class ProgressEvent {
  constructor(data) {
    this._data = data;
  }

  get type() { return this._data.type || ''; }
  get percent() { return this._data.percent || 0; }
  get speed() { return this._data.speed || ''; }
  get eta() { return this._data.eta || ''; }
  get path() { return this._data.path || null; }
  get size() { return this._data.size || null; }
  get message() { return this._data.message || ''; }
}

class DownloadRequest {
  constructor(dw, url, mediaType = null) {
    this._dw = dw;
    this._url = url;
    this._mediaType = mediaType;
    this._overrides = {};
    this._onProgress = null;
    this._onComplete = null;
    this._onError = null;
  }

  format(fmt) {
    this._overrides.format = fmt;
    return this;
  }

  quality(q) {
    this._overrides.quality = q;
    return this;
  }

  output(path) {
    this._overrides.output = path;
    return this;
  }

  parallel(n) {
    this._overrides.parallel = n;
    return this;
  }

  embedMetadata(enabled) {
    this._overrides.embedMetadata = enabled;
    return this;
  }

  embedThumbnail(enabled) {
    this._overrides.embedThumbnail = enabled;
    return this;
  }

  embedSubtitles(enabled) {
    this._overrides.embedSubtitles = enabled;
    return this;
  }

  onProgress(callback) {
    this._onProgress = callback;
    return this;
  }

  onComplete(callback) {
    this._onComplete = callback;
    return this;
  }

  onError(callback) {
    this._onError = callback;
    return this;
  }

  run() {
    return new Promise((resolve, reject) => {
      const config = { ...this._dw.config, ...this._overrides };
      const args = ['download', this._url, '--json'];

      if (config.format) args.push('--format', config.format);
      if (config.quality) args.push('--quality', config.quality);
      if (config.output) args.push('--output', config.output);
      if (config.parallel) args.push('--parallel', String(config.parallel));
      if (config.embedMetadata) args.push('--embed-metadata');
      if (config.embedThumbnail) args.push('--embed-thumbnail');
      if (config.embedSubtitles) args.push('--embed-subtitles');
      if (this._mediaType === 'audio') args.push('--audio-only');
      if (this._mediaType === 'video') args.push('--video-only');

      const proc = spawn('souris-dw', args);
      let result = {};

      proc.stdout.on('data', (data) => {
        const lines = data.toString().split('\n').filter(l => l.trim());
        for (const line of lines) {
          try {
            const event = JSON.parse(line);
            const progress = new ProgressEvent(event);

            if (event.type === 'progress' && this._onProgress) {
              this._onProgress(progress);
            } else if (event.type === 'complete') {
              result = event;
              if (this._onComplete) this._onComplete(event);
            } else if (event.type === 'error') {
              const msg = event.message || 'Unknown error';
              if (this._onError) this._onError(msg);
              reject(new Error(msg));
              return;
            }
          } catch (e) {
            // Skip invalid JSON
          }
        }
      });

      proc.stderr.on('data', (data) => {
        // Collect stderr for error messages
      });

      proc.on('close', (code) => {
        if (code !== 0 && !result.type) {
          reject(new Error(`Download failed with code ${code}`));
        } else {
          resolve(result);
        }
      });
    });
  }
}

module.exports = { DownloadRequest, ProgressEvent };

File: sdks/go/request.go
package sourisdw

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os/exec"
)

// ProgressEvent represents a download progress event.
type ProgressEvent struct {
	Type    string  `json:"type"`
	Percent float64 `json:"percent"`
	Speed   string  `json:"speed"`
	Eta     string  `json:"eta"`
	Path    string  `json:"path,omitempty"`
	Size    int64   `json:"size,omitempty"`
	Message string  `json:"message,omitempty"`
}

// DownloadRequest is a fluent download request.
type DownloadRequest struct {
	dw          *SourisDW
	url         string
	mediaType   string
	overrides   map[string]interface{}
	onProgress  func(ProgressEvent)
	onComplete  func(map[string]interface{})
	onError     func(string)
}

// Format overrides the format.
func (r *DownloadRequest) Format(fmt string) *DownloadRequest {
	r.setOverride("format", fmt)
	return r
}

// Quality overrides the quality.
func (r *DownloadRequest) Quality(q string) *DownloadRequest {
	r.setOverride("quality", q)
	return r
}

// Output overrides the output directory.
func (r *DownloadRequest) Output(path string) *DownloadRequest {
	r.setOverride("output", path)
	return r
}

// Parallel overrides the parallel downloads.
func (r *DownloadRequest) Parallel(n int) *DownloadRequest {
	r.setOverride("parallel", n)
	return r
}

// EmbedMetadata overrides metadata embedding.
func (r *DownloadRequest) EmbedMetadata(enabled bool) *DownloadRequest {
	r.setOverride("embed_metadata", enabled)
	return r
}

// EmbedThumbnail overrides thumbnail embedding.
func (r *DownloadRequest) EmbedThumbnail(enabled bool) *DownloadRequest {
	r.setOverride("embed_thumbnail", enabled)
	return r
}

// EmbedSubtitles overrides subtitle embedding.
func (r *DownloadRequest) EmbedSubtitles(enabled bool) *DownloadRequest {
	r.setOverride("embed_subtitles", enabled)
	return r
}

// OnProgress sets the progress callback.
func (r *DownloadRequest) OnProgress(cb func(ProgressEvent)) *DownloadRequest {
	r.onProgress = cb
	return r
}

// OnComplete sets the completion callback.
func (r *DownloadRequest) OnComplete(cb func(map[string]interface{})) *DownloadRequest {
	r.onComplete = cb
	return r
}

// OnError sets the error callback.
func (r *DownloadRequest) OnError(cb func(string)) *DownloadRequest {
	r.onError = cb
	return r
}

// Run executes the download.
func (r *DownloadRequest) Run() (map[string]interface{}, error) {
	args := []string{"download", r.url, "--json"}

	config := r.dw.config
	if v, ok := r.overrides["format"]; ok {
		args = append(args, "--format", v.(string))
	} else if config.Format != "" {
		args = append(args, "--format", config.Format)
	}

	if v, ok := r.overrides["quality"]; ok {
		args = append(args, "--quality", v.(string))
	} else if config.Quality != "" {
		args = append(args, "--quality", config.Quality)
	}

	if v, ok := r.overrides["output"]; ok {
		args = append(args, "--output", v.(string))
	} else if config.Output != "" {
		args = append(args, "--output", config.Output)
	}

	if v, ok := r.overrides["parallel"]; ok {
		args = append(args, "--parallel", fmt.Sprintf("%d", v.(int)))
	} else if config.Parallel > 0 {
		args = append(args, "--parallel", fmt.Sprintf("%d", config.Parallel))
	}

	if config.EmbedMetadata {
		args = append(args, "--embed-metadata")
	}

	if config.EmbedThumbnail {
		args = append(args, "--embed-thumbnail")
	}

	if r.mediaType == "audio" {
		args = append(args, "--audio-only")
	} else if r.mediaType == "video" {
		args = append(args, "--video-only")
	}

	cmd := exec.Command("souris-dw", args...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("pipe failed: %w", err)
	}

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("start failed: %w", err)
	}

	result := make(map[string]interface{})
	scanner := bufio.NewScanner(stdout)

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}

		var event map[string]interface{}
		if err := json.Unmarshal([]byte(line), &event); err != nil {
			continue
		}

		eventType, _ := event["type"].(string)

		switch eventType {
		case "progress":
			if r.onProgress != nil {
				pe := ProgressEvent{
					Type:    eventType,
					Percent: event["percent"].(float64),
					Speed:   event["speed"].(string),
					Eta:     event["eta"].(string),
				}
				r.onProgress(pe)
			}
		case "complete":
			result = event
			if r.onComplete != nil {
				r.onComplete(event)
			}
		case "error":
			msg, _ := event["message"].(string)
			if r.onError != nil {
				r.onError(msg)
			}
			return nil, fmt.Errorf("download error: %s", msg)
		}
	}

	if err := cmd.Wait(); err != nil {
		return nil, fmt.Errorf("download failed: %w", err)
	}

	return result, nil
}

func (r *DownloadRequest) setOverride(key string, value interface{}) {
	if r.overrides == nil {
		r.overrides = make(map[string]interface{})
	}
	r.overrides[key] = value
}
